# Примеры и упражнения по Python3  - средний уровень

Автор-составитель - Михаил Колодин

Версия от 2021-04-16 1.0

Разделы:
* Параметры системы
* Циклы
* Строки
* Кортежи и списки
* Словари и множества
* Прочие сложные структуры данных
* Функции
* Математика

**Параметры системы**

In [105]:
import sys
print("Python version:", sys.version)

from platform import python_version
print(python_version())

Python version: 3.8.3 (default, Jul  2 2020, 16:21:59) 
[GCC 7.3.0]
3.8.3


**Циклы**

***Факториал $n! = 1 \cdot 2 \cdot 3 \cdot \dots \cdot n$***

In [2]:
UP = 30
f = 1

In [3]:
for i in range(1, UP):
    f *= i             #  то же, что f = f * i
    print("%10i!  =" % i, "%40i" % f)

         1!  =                                        1
         2!  =                                        2
         3!  =                                        6
         4!  =                                       24
         5!  =                                      120
         6!  =                                      720
         7!  =                                     5040
         8!  =                                    40320
         9!  =                                   362880
        10!  =                                  3628800
        11!  =                                 39916800
        12!  =                                479001600
        13!  =                               6227020800
        14!  =                              87178291200
        15!  =                            1307674368000
        16!  =                           20922789888000
        17!  =                          355687428096000
        18!  =                         640237370

***Константа е - аппроксимация***

См. https://ru.wikipedia.org/wiki/E_(%D1%87%D0%B8%D1%81%D0%BB%D0%BE) 

$e = \sum\limits_{n=0}^\infty \frac1{n!} = \frac1{1!} + \frac1{2!} + \frac1{3!} + \frac1{4!} + \frac1{5!} + ... = \frac11 + \frac12 + \frac16 + \frac1{24} + \frac1{120} + ...$

In [34]:
UP = 30
add = 1
sum = 1

In [35]:
for i in range(1, UP):
    add /= i
    sum += add
    print("%30.25g -> %30.25g" % (add, sum))

                             1 ->                              2
                           0.5 ->                            2.5
   0.1666666666666666574148081 ->      2.66666666666666651863693
  0.04166666666666666435370203 ->      2.70833333333333303727386
 0.008333333333333333217685102 ->     2.716666666666666341001246
 0.001388888888888888941894328 ->     2.718055555555555447000415
0.0001984126984126984125263171 ->     2.718253968253968366752815
2.480158730158730156578964e-05 ->     2.718278769841270037233016
2.755731922398589251095059e-06 ->     2.718281525573192247691168
2.755731922398589356974178e-07 ->     2.718281801146384513145904
2.505210838544172022386618e-08 ->     2.718281826198492900914516
2.087675698786810018655515e-09 ->     2.718281828286168710917536
1.605904383682161592580233e-10 ->     2.718281828446759362805096
 1.14707455977297261228837e-11 ->     2.718281828458230187095523
7.647163731819817415255803e-13 ->     2.718281828458994908714885
4.77947733238738588453487

**Строки**

***Шифр Цезаря***

In [65]:
# setup shift
shift = 1
# setup method
import string
alfa = string.digits + string.ascii_letters
alfa_len = len(alfa)
print(alfa_len, alfa)

62 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


In [66]:
sopen = "Hello commander! Start at noon! Ceasar"

#sopen = "Здравствуй, Цезарь! Идущие на смерть тебя приветствуют!"
#sopen = "Ave, Caesar, morituri te salutant"
#sopen = "Николай неохотно поплёлся к директору: «Ave, Caesar, morituri te salutant!» — воскликнул он в дверях."

shift = 1
print(sopen)

Hello commander! Start at noon! Ceasar


In [67]:
scoded = "".join(alfa[alfa.index(c) + shift] if c in alfa else c for c in sopen)
print(scoded)

Ifmmp dpnnboefs! Tubsu bu oppo! Dfbtbs


In [68]:
sback = "".join(alfa[alfa.index(c) - shift] if c in alfa else c for c in scoded)
print(sback)

Hello commander! Start at noon! Ceasar


***Подсчёт символов***

(найти 5 самых частых букв)

In [136]:
#stroka = "twas brillig and the slighty towes"
stroka = "A Counter is a container that keeps track of how many times equivalent values are added. It can be used to implement the same algorithms for which bag or multiset data structures are commonly used in other languages."\
    "Counter supports three forms of initialization. Its constructor can be called with a sequence of items, a dictionary containing keys and counts, or using keyword arguments mapping string names to counts."\
    "This example counts the letters appearing in all of the words in the system dictionary to produce a frequency distribution, then prints the three most common letters. Leaving out the argument to most_common() produces a list of all the items, in order of frequency."

print(stroka)

A Counter is a container that keeps track of how many times equivalent values are added. It can be used to implement the same algorithms for which bag or multiset data structures are commonly used in other languages.Counter supports three forms of initialization. Its constructor can be called with a sequence of items, a dictionary containing keys and counts, or using keyword arguments mapping string names to counts.This example counts the letters appearing in all of the words in the system dictionary to produce a frequency distribution, then prints the three most common letters. Leaving out the argument to most_common() produces a list of all the items, in order of frequency.


In [143]:
# не по-питоновски (ну, частично)
cnt = {}

# counting all characters
for c in stroka:
    if c not in cnt:
        cnt[c] = 0
    cnt[c] += 1
        
# finding max by sorting
arr = [(k, v) for k, v in cnt.items()]
print("unsorted:", arr)

arr.sort(key = lambda x: x[1], reverse = True)
print("sorted:", arr)

loc = [x[0] for x in arr[:5]]

print("\n* most often characters are:", loc)

unsorted: [('A', 1), (' ', 110), ('C', 2), ('o', 45), ('u', 25), ('n', 43), ('t', 61), ('e', 62), ('r', 37), ('i', 37), ('s', 40), ('a', 40), ('c', 23), ('h', 18), ('k', 4), ('p', 12), ('f', 10), ('w', 5), ('m', 24), ('y', 9), ('q', 4), ('v', 3), ('l', 18), ('d', 16), ('.', 6), ('I', 2), ('b', 4), ('g', 12), ('z', 1), (',', 4), ('T', 1), ('x', 1), ('L', 1), ('_', 1), ('(', 1), (')', 1)]
sorted: [(' ', 110), ('e', 62), ('t', 61), ('o', 45), ('n', 43), ('s', 40), ('a', 40), ('r', 37), ('i', 37), ('u', 25), ('m', 24), ('c', 23), ('h', 18), ('l', 18), ('d', 16), ('p', 12), ('g', 12), ('f', 10), ('y', 9), ('.', 6), ('w', 5), ('k', 4), ('q', 4), ('b', 4), (',', 4), ('v', 3), ('C', 2), ('I', 2), ('A', 1), ('z', 1), ('T', 1), ('x', 1), ('L', 1), ('_', 1), ('(', 1), (')', 1)]

* most often characters are: [' ', 'e', 't', 'o', 'n']


In [134]:
# по-питоновски
import collections as coll

cnt = coll.Counter(stroka)
print("* most often characters are:", [x[0] for x in cnt.most_common(5)])

* most often characters are: [' ', 't', 's', 'i', 'l']


**Кортежи и списки**

**Словари и множества**

In [140]:
# TODO
pass

**Прочие сложные структуры данных**

In [141]:
# TODO
pass

**Функции**

In [142]:
# TODO
pass

***Генератор речей***

In [50]:
# mk-genspeech.py 2014-05-25 1.1
# Генератор речей
# Mikhail (myke) Kolodin, mykespb@gmail.com 

import random

text = """
  Товарищи! / Господа! / Коллеги! / Дорогие товарищи! / Уважаемые коллеги! / Дамы и господа! /  Уважаемые участники! / Друзья! / Граждане соучастники! / Граждане и гражданки! / / / / / / /
  С другой стороны / Равным образом / Не следует, однако, забывать о том, что /  Таким образом / Повседневная практика показывает. что / Значимость этих проблем настолько очевидна, что / Из выступление товарища начальника ясно следует, что / В решениях съезда нашей партии указано, что/ Разнообразный и богатый опыт / Задача организации, в особенности же / Идейные соображения  высшего порядка, а также / Принимая во внимание, что / Вы, конечно, знаете, что / Вам, должно быть, известен тот факт, что
  реализация намеченных плановых заданий / рамки и место обучения кадров / постоянный количественный рост и сфера нашей активности / сложившаяся структура организации / новая модель организационной деятельности / дальнейшее развитие различных форм деятельности / постоянное информационно-пропагандистское обеспечение нашей деятельности / управление и развитие структуры / консультация с широким активом / начало повседневной работы по формированию позиции
  играет важную роль в формировании / требуют от нас анализа / требуют определения и уточнения /  способствуют подготовке и реализации / обеспечивают широкому кругу специалистов участие в формировании
  , что позволяет выполнить важные задания по разработке / , что в значительной степени обуславливает создание / , что позволяет оценить значение / , что представляет собой интересный эксперимент / , что влечёт за собой интересный процесс внедрения и модернизации
  существенных финансовых и административных условий / дальнейших направлений развития /  системы массового участия / позиций, занимаемых участниками в отношении поставленных задач /  новых предложений / направлений прогрессивного развития / форм воздействия /  модели развития / соответствующих условий активизации / системы обучения кадров, соответствующей  насущным потребностям
  """

def make_speech (rows=1):
    parts = [part.split('/') for part in text.split('\n')]
    for num in range(1, rows+1):
        print ("Пункт %d.\n" % num + "".join([random.choice(part) for part in parts]) + '.')

make_speech(4)

Пункт 1.
 Дамы и господа!  Задача организации, в особенности же  новая модель организационной деятельности  требуют определения и уточнения  , что представляет собой интересный эксперимент  соответствующих условий активизации   .
Пункт 2.
  Товарищи!  Принимая во внимание, что  начало повседневной работы по формированию позиции требуют определения и уточнения  , что позволяет оценить значение  системы обучения кадров, соответствующей  насущным потребностям  .
Пункт 3.
   Таким образом  сложившаяся структура организации  требуют от нас анализа  , что в значительной степени обуславливает создание  позиций, занимаемых участниками в отношении поставленных задач   .
Пункт 4.
 Задача организации, в особенности же  постоянный количественный рост и сфера нашей активности  требуют от нас анализа  , что влечёт за собой интересный процесс внедрения и модернизации форм воздействия   .


**Математика**

***Простые числа***

In [12]:
# простые числа (primes)
def primes (lim=100):
    """ show primes up to lim """
    found = []
    for tested in range (2, lim+1):
        for tester in found:
            if tested % tester == 0:
                break
        else:
            found.append (tested)
            print (tested, end=" ")

primes(1000)

2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 101 103 107 109 113 127 131 137 139 149 151 157 163 167 173 179 181 191 193 197 199 211 223 227 229 233 239 241 251 257 263 269 271 277 281 283 293 307 311 313 317 331 337 347 349 353 359 367 373 379 383 389 397 401 409 419 421 431 433 439 443 449 457 461 463 467 479 487 491 499 503 509 521 523 541 547 557 563 569 571 577 587 593 599 601 607 613 617 619 631 641 643 647 653 659 661 673 677 683 691 701 709 719 727 733 739 743 751 757 761 769 773 787 797 809 811 821 823 827 829 839 853 857 859 863 877 881 883 887 907 911 919 929 937 941 947 953 967 971 977 983 991 997 

***Линейное уравнение***

In [16]:
# дано: ax + b = 0
a = -3
b = 6
print(f"линейное уравнение {a}x{b:+n} = 0")

линейное уравнение -3x+6 = 0


In [17]:
# решение: x = -b / a
# если a!= 0
if a:
    x = -b / a
    print(f"решение: x = {x}")
else:
    print("а=0, это не линейное уравнение!")

решение: x = 2.0


***Квадратное уравнение***

Дано: $ax^2 + bx + c = 0$, найти $x$

Решение: $x_{1, 2} = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$, либо $x = \frac{-b}{2a}$

In [37]:
import math

In [69]:
# дано: a, b, c
a = 1
b = 6
c = -8
print(f"Решаем квадратное уравнение с коэффициентами {a=}, {b=}, {c=}")

Решаем квадратное уравнение с коэффициентами a=1, b=6, c=-8


In [70]:
if a:
    
    d = b**2 - 4*a*c      # это дискриминант
    print(f"дискриминант равен {d},", 
     "положительный (2 решения)" if d>0 else
     "отрицательный (нет решений)" if d<0 else
     "нулевой (1 решение)" )

    if d > 0:
        x1 = (-b + math.sqrt(d)) / (2*a)
        x2 = (-b - math.sqrt(d)) / (2*a)
        print(f"2 решения:\nx1 = {x1}\nx2 = {x2}")
    elif d == 0:
        x = -b / (2*a)
        print(f"1 решение:\nx = {x}")
    else:
        print("решений в R нет")

else:
    print("a=0, это не квадратное уравнение")

дискриминант равен 68, положительный (2 решения)
2 решения:
x1 = 1.1231056256176606
x2 = -7.123105625617661
